tests/unit/test_business_logic.py

```diff
@@ -2267,110 +2340,122 @@ def get_comments(user):
+    assert not visible.vote(RequestFileVote.REJECTED, checkers[0])
+    assert not visible.comment("turn 3 checker 1 private", checkers[1])
+    assert not visible.comment("turn 3 checker 0 public", checkers[0])
 
     bll.return_request(release_request, checkers[0])
+    release_request = factories.refresh_release_request(release_request)
+    assert release_request.review_turn == 4
 
-    # in RequestPhase.COMPLETE for a 2nd round
-    # Checkers should see all historic comments,
+    # in ReviewTurnPhase.COMPLETE for a 2nd round
     # Author should see previous round's public comments, but not any private comments
-    assert get_comments(checkers[0]) == [
-        "turn 1 checker 0 private",
-        "turn 1 checker 1 private",
-        "turn 1 checker 0 public",
-        "turn 2 author public",
-        "turn 3 checker 0 private",
-        "turn 3 checker 1 private",
-        "turn 3 checker 0 public",
-    ]
-    assert get_comments(checkers[1]) == [
-        "turn 1 checker 0 private",
-        "turn 1 checker 1 private",
-        "turn 1 checker 0 public",
-        "turn 2 author public",
-        "turn 3 checker 0 private",
-        "turn 3 checker 1 private",
-        "turn 3 checker 0 public",
-    ]
-    assert get_comments(author) == [
-        "turn 1 checker 0 public",
-        "turn 2 author public",
-        "turn 3 checker 0 public",
-    ]
-
-    assert release_request.review_turn == 4
+    for checker in checkers:
+        visible = get_visible_items(checker)
+        assert visible.comment("turn 1 checker 0 private", checkers[0])
+        assert visible.vote(RequestFileVote.APPROVED, checkers[0])
+        assert visible.comment("turn 1 checker 1 private", checkers[1])
+        assert visible.vote(RequestFileVote.APPROVED, checkers[1])
+        assert visible.comment("turn 1 checker 0 public", checkers[0])
+        assert visible.comment("turn 2 author public", author)
+        assert visible.comment("turn 3 checker 0 private", checkers[0])
+        assert visible.vote(RequestFileVote.REJECTED, checkers[0])
+        assert visible.comment("turn 3 checker 1 private", checkers[1])
+        assert visible.comment("turn 3 checker 0 public", checkers[0])
+
+    # author sees no private comments, and no turn 3 things.
+    visible = get_visible_items(author)
+    assert not visible.comment("turn 1 checker 0 private", checkers[0])
+    assert visible.vote(RequestFileVote.APPROVED, checkers[0])
+    assert not visible.comment("turn 1 checker 1 private", checkers[1])
+    assert visible.vote(RequestFileVote.APPROVED, checkers[1])
+    assert visible.comment("turn 1 checker 0 public", checkers[0])
+    assert visible.comment("turn 2 author public", author)
+    assert not visible.comment("turn 3 checker 0 private", checkers[0])
+    assert visible.vote(RequestFileVote.REJECTED, checkers[0])
+    assert not visible.comment("turn 3 checker 1 private", checkers[1])
+    assert visible.comment("turn 3 checker 0 public", checkers[0])
```

Comment:
This is missing a test of comments in a turn before the request moves into a final state.  i.e. comments made just before a request is rejected/released.

Reply:
Good catch, it is, which is why we've not caught the bug about rejected requests visibility.